### PR TITLE
[leaflet] Making the GeoJSON methods accessible as static methods

### PR DESCRIPTION
**Pull request by reblace**
Please fill in this template.

- [X] Make your PR against the `master` branch.
- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped#common-mistakes).
- [X] Run `tsc` without errors.

If changing an existing definition:
- [X] Provide a URL to documentation or source code which provides context for the suggested changes: http://leafletjs.com/reference-1.0.2.html

The primary thing I change in this PR is to move the static methods on the GeoJSON object from being declared on the GeoJSON interface to being declared as static methods on the GeoJSON class. The reason for this change is that the typing file only provides access to the GeoJSON interface through the factory methods used to create GeoJSON layers. The factory method returns an object that doesn't actually provide the static methods. So, you can currently write code that will compile against the typing file but encounter a runtime error because the static methods are undefined.

And, you can't currently compile if you try to access the static methods through the L.GeoJSON class. By adding the static methods to the L.GeoJSON class, you can now access the methods correctly and compile and they work at runtime.

Other smaller fixes had to do with the fact that two of the static methods take as parameters or return arbitrarily nested arrays. The typings were limiting these unnecessarily. While opening them up to any[] is less helpful when writing code against the API, it makes more sense than limiting their utility.

### CODE REVIEW

File: leaflet/index.d.ts [hunk 1]
```diff
@@ -817,67 +817,72 @@ declare namespace L {
...
+        static coordsToLatLngs(
+            coords: any[],
+            levelsDeep?: number,
+            coordsToLatLng?: (coords: [number, number] | [number, number, number]) => LatLng): any[]; // Using any[] to avoid artificially limiting valid calls
 
         /**
          * Reverse of coordsToLatLng
          */
-        latLngToCoords(latlng: LatLng): [number, number] | [number, number, number];
+        static latLngToCoords(latlng: LatLng): number[];
```
Comment (alejo90) [inline leaflet/index.d.ts]: Why the change in the return type? According to the docs, the only options are arrays with two or three elements.

File: leaflet/index.d.ts [hunk 2]
```diff
@@ -817,67 +817,72 @@ declare namespace L {
...
+
+        static asFeature(geojson: GeoJSON.Feature<GeoJSON.GeometryObject>): GeoJSON.Feature<GeoJSON.GeometryObject>;
+
+    }
+
+    /**
+     * Represents a GeoJSON object or an array of GeoJSON objects.
+     * Allows you to parse GeoJSON data and display it on the map. Extends FeatureGroup.
+     */
+    export interface GeoJSON extends FeatureGroup {
```
Comment (alejo90) [inline leaflet/index.d.ts]: Could you get rid of this interface and put the methods in the class, please.

**Review by alejo90 — CHANGES_REQUESTED** [review]
Thanks for your help. We just need a couple of changes before merging.

File: leaflet/index.d.ts (hunk 1 above)
Comment (reblace) [inline leaflet/index.d.ts]: With the old definition, the following code gives a compiler error:

```js
let threeCoords: [number, number, number] = [1, 2, 3];
latLng = L.GeoJSON.coordsToLatLng(threeCoords);
threeCoords = L.GeoJSON.latLngToCoords(latLng);
```

With this error:
error TS2322: Type '[number, number] | [number, number, number]' is not assignable to type '[number, number, number]'.
  Type '[number, number]' is not assignable to type '[number, number, number]'.
    Property '2' is missing in type '[number, number]'.

The typescript docs (https://www.typescriptlang.org/docs/handbook/advanced-types.html) state that "If a value has the type A | B, we only know for certain that it has members that both A and B have." which means that we can only assign the return value of the 'coordsToLatLng' function to [number, number] because the the [number, number, number] tuple has a property '2' (the third value in the tuple) that doesn't exist on the [number, number] tuple. The union type works ok in the parameter definition because the constraint is on the consumer of the type, not the provider of the type.

Does that make sense? I'm not an expert in Typescript, but I ran into the issue when I added the GeoJSON stuff to the tests and then found that documentation on the Typescript docs site.

Comment (reblace) [inline leaflet/index.d.ts]: Actually, it seems I can make the compile happy while still typing this tightly if I use overloading and order them in the right order. I'll update the PR and you can tell me what you think.

Comment (alejo90) [inline leaflet/index.d.ts]: Sounds good, thanks.

File: leaflet/index.d.ts [hunk 3]
```diff
@@ -817,67 +817,73 @@ declare namespace L {
...
+        static coordsToLatLngs(
+            coords: any[],
+            levelsDeep?: number,
+            coordsToLatLng?: (coords: [number, number] | [number, number, number]) => LatLng): any[]; // Using any[] to avoid artificially limiting valid calls
 
         /**
          * Reverse of coordsToLatLng
          */
-        latLngToCoords(latlng: LatLng): [number, number] | [number, number, number];
+        static latLngToCoords(latlng: LatLng): [number, number, number]; // A three tuple can be assigned to a two or three tuple
```
Comment (reblace) [inline leaflet/index.d.ts]: This is the best compromise imo. We can't/shouldn't use union types here for the reasons I described in the main comment thread. A three tuple is assignable to a one, two, or three tuple. But a three tuple cannot be assigned to a four (or greater) tuple. Which is more restrictive than just saying it's a number array.

File: leaflet/index.d.ts (hunk 2 above)
Comment (reblace) [inline leaflet/index.d.ts]: BLUF: We can't easily get rid of the GeoJSON interface. And, leaving them separate actually provides exactly the functionality we want in this context.


**Long Version:**

Why it would be hard:
To get rid of the GeoJSON interface, we'd need to move the remaining methods over to the class. But, then we'd also need to make sure the GeoJSON class was implementing the FeatureGroup interface. But since there are no other declared classes, we can't just have GeoJSON extend FeatureGroup, we'd have to actually add all the expected methods to the GeoJSON class.

Leaflet uses factories for everything, so the typings generally don't need to define any constructors or classes. In this case, we needed to add the class so we could add static methods (since you can't declare static methods on an interface). Since we don't add a public constructor on the class, you still can't create a GeoJSON instance apart from the factory methods. But now you can call the static methods on the class.

Why this approach is ok (I think):
In TS it seems that interface and class definitions of the same name are merged by the compiler. This means that the 'GeoJSON' type that is returned by the factory method that creates a GeoJSON object is returning the union of the GeoJSON class and interface methods. In our case specifically the only impact that the merge has is to make it so that we can reference the type "GeoJSON" with all of the interface methods, while also directly invoking static methods on the "GeoJSON" class (eg. GeoJSON.doStaticStuff()) - which is exactly the behavior we want.

Let me know if you have an alternative approach you'd prefer.

Comment (alejo90) [inline leaflet/index.d.ts]: Ok, I guess we can leave that for later since it will take more effort to validate.

**Comment by alejo90** [comment]
👍

**Comment by horiuchi** [comment]
👍
